hal/halx86/acpi/madt.c

```diff
@@ -12,77 +13,250 @@
+#endif // _MSC_VER
+#endif // DBG && 0
+
 /* GLOBALS ********************************************************************/
 
-PROCESSOR_IDENTITY HalpStaticProcessorIdentity[MAXIMUM_PROCESSORS] = {{0}};
-PPROCESSOR_IDENTITY HalpProcessorIdentity = NULL;
 HALP_APIC_INFO_TABLE HalpApicInfoTable;
-ACPI_TABLE_MADT *MadtTable;
-ACPI_SUBTABLE_HEADER *AcpiHeader;
-ACPI_MADT_LOCAL_APIC *LocalApic;
+
+// ACPI_MADT_LOCAL_APIC.LapicFlags masks
+#define LAPIC_FLAG_ENABLED          0x00000001
+#define LAPIC_FLAG_ONLINE_CAPABLE   0x00000002
+// Bits 2-31 are reserved.
+
+static PROCESSOR_IDENTITY HalpStaticProcessorIdentity[MAXIMUM_PROCESSORS];
+const PPROCESSOR_IDENTITY HalpProcessorIdentity = HalpStaticProcessorIdentity;
+
+#if 0
+extern ULONG HalpPicVectorRedirect[16];
+#endif
 
 /* FUNCTIONS ******************************************************************/
 
+// Note: HalpParseApicTables() is called early, so its DPRINT*() do nothing.
 VOID
 HalpParseApicTables(
     _In_ PLOADER_PARAMETER_BLOCK LoaderBlock)
 {
+    ACPI_TABLE_MADT *MadtTable;
+    ACPI_SUBTABLE_HEADER *AcpiHeader;
     ULONG_PTR TableEnd;
-    ULONG ValidProcessorCount;
 
-    /* We only support legacy APIC for now, this will be updated in the future */
-    HalpApicInfoTable.ApicMode = 0x10;
-    MadtTable = HalAcpiGetTable(LoaderBlock, 'CIPA');
+    MadtTable = HalAcpiGetTable(LoaderBlock, APIC_SIGNATURE);
+    if (!MadtTable)
+    {
+        DPRINT01("HalAcpiGetTable() failed\n");
+        return;
+    }
+
+    if (MadtTable->Header.Length < sizeof(*MadtTable))
+    {
+        DPRINT01("Length is too short: %p, %u\n", MadtTable, MadtTable->Header.Length);
+        return;
+    }
 
-    AcpiHeader = (ACPI_SUBTABLE_HEADER*)MadtTable;
-    AcpiHeader->Length = sizeof(ACPI_TABLE_MADT);
-    TableEnd = (ULONG_PTR)MadtTable + MadtTable->Header.Length;
+    DPRINT00("MADT table: Address %08X, Flags %08X\n", MadtTable->Address, MadtTable->Flags);
 
-    HalpApicInfoTable.ProcessorCount = 0;
-    HalpProcessorIdentity = HalpStaticProcessorIdentity;
+#if 1
 
-    AcpiHeader = (ACPI_SUBTABLE_HEADER*)MadtTable;
-    AcpiHeader->Length = sizeof(ACPI_TABLE_MADT);
-    TableEnd = (ULONG_PTR)MadtTable + MadtTable->Header.Length;
+    // TODO: We support only legacy APIC for now
+    HalpApicInfoTable.ApicMode = HALP_APIC_MODE_LEGACY;
+    // TODO: What about 'MadtTable->Flags & ACPI_MADT_PCAT_COMPAT'?
 
-    while ((ULONG_PTR)AcpiHeader <= TableEnd)
+#else // TODO: Is that correct?
+
+    if ((MadtTable->Flags & ACPI_MADT_PCAT_COMPAT) == ACPI_MADT_DUAL_PIC)
+    {
+        HalpApicInfoTable.ApicMode = HALP_APIC_MODE_LEGACY;
+    }
+    else // if ((MadtTable->Flags & ACPI_MADT_PCAT_COMPAT) == ACPI_MADT_MULTIPLE_APIC)
     {
-        LocalApic = (ACPI_MADT_LOCAL_APIC*)AcpiHeader;
+#if 1
+        DPRINT01("ACPI_MADT_MULTIPLE_APIC support is UNIMPLEMENTED\n");
+        return;
+#else
+        HalpApicInfoTable.ApicMode = HALP_APIC_MODE_xyz;
```

Comment:
What's this for?

Reply:
It's the best suggestion I can come up with (see TODOs at new line 83 and previous): a placeholder for further development.